engine/src/main/java/org/hibernate/search/engine/search/common/Rewrite.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.engine.search.common;
+
+/**
+ * The rewrite parameter determines:
+ *
+ * <ul>
+ *     <li>How Lucene calculates the relevance scores for each matching document</li>
+ *     <li>Whether Lucene changes the original query to a bool query or bit set</li>
+ *     <li>If changed to a bool query, which term query clauses are included</li>
+ * </ul>
+ */
+public enum Rewrite {
+
+	/**
+	 * Assigns each document a relevance score equal to the boost parameter
+	 */
+	CONSTANT_SCORE_BLENDED,
+	/**
+	 * Uses the {@link Rewrite#CONSTANT_SCORE_BOOLEAN} method for fewer matching terms.
+	 * Otherwise, this method finds all matching terms in sequence and returns matching documents using a bit set.
+	 */
+	CONSTANT_SCORE,
+	/**
+	 * Assigns each document a relevance score equal to the boost parameter.
+	 */
+	CONSTANT_SCORE_BOOLEAN,
+	/**
+	 * Calculates a relevance score for each matching document.
+	 */
+	SCORING_BOOLEAN,
+	/**
+	 * Calculates a relevance score for each matching document as if all terms had the same frequency.
+	 * This frequency is the maximum frequency of all matching terms.
+	 */
+	TOP_TERMS_BLENDED_FREQS_N,
+	/**
+	 * Assigns each matching document a relevance score equal to the boost parameter.
+	 */
+	TOP_TERMS_BOOST_N,
+	/**
+	 * Calculates a relevance score for each matching document.
+	 */
+	TOP_TERMS_N;
```

Comment:
My, that's complicated.

Two suggestions:

* You might want to check those names are consistent across Lucene/Elasticsearch/OpenSearch, and if not, pick the most self-explanatory set of names.
* Maybe add some info to the Javadoc regarding use cases for each rewrite method? Or at least to the reference documentation.

It would probably help